addons/mail/static/src/core/common/thread_model.js

```diff
@@ -834,26 +837,37 @@ export class Thread extends Record {
      * @param {import("models").Message} [messageId] if not provided, load around newest message
      */
     async loadAround(messageId) {
-        if (!this.messages.some(({ id }) => id === messageId)) {
+        if (
+            this.lastLoadAroundId === messageId ||
+            (this.isLoaded && this.messages.some(({ id }) => id === messageId))
+        ) {
+            return;
+        }
+        this.lastLoadAroundId = messageId;
+        try {
             this.isLoaded = false;
             this.scrollTop = undefined;
-            const { messages } = await rpc(this.getFetchRoute(), {
-                ...this.getFetchParams(),
+            this.messages = await this.fetchMessages({
                 around: messageId,
+                limit: this.store.FETCH_LIMIT * 2,
             });
             this.isLoaded = true;
-            this.messages = this.store.Message.insert(messages.reverse(), { html: true });
-            this.loadNewer = messageId ? true : false;
+            this.loadNewer = messageId !== undefined ? true : false;
             this.loadOlder = true;
-            if (messages.length < this.store.FETCH_LIMIT) {
-                const olderMessagesCount = messages.filter(({ id }) => id < messageId).length;
-                if (olderMessagesCount < this.store.FETCH_LIMIT / 2) {
+            if (this.messages.length < this.store.FETCH_LIMIT * 2) {
+                const olderMessagesCount = this.messages.filter(({ id }) => id < messageId).length;
+                const newerMessagesCount = this.messages.filter(({ id }) => id > messageId).length;
+                if (olderMessagesCount < this.store.FETCH_LIMIT - 1) {
                     this.loadOlder = false;
-                } else {
+                }
+                if (newerMessagesCount < this.store.FETCH_LIMIT) {
                     this.loadNewer = false;
                 }
             }
             this._enrichMessagesWithTransient();
+        } catch {
+            this.lastLoadAroundId = null;
+            // handled in fetchMessages
```

Comment:
`lastLoadAroundId` is prone to errors, because this works only when `loadAround` is the main call, which is not true as we have `fetchMore`/`fetchNew`. This is not a practical issue because these functions do not remove messages... But there's plain to `unloadMessages` as there's a performance issue to have too many messages loaded.

All in all, this function would be much easier to maintain if it is coded to work when being used all the time or barely. What about just a field `isLoadingAround` that is `true` during it and then there's `.finally { isLoadingAround = false }`?